### PR TITLE
LIBXSMM in TensorFlow broken due to TensorFlow API changes

### PR DESCRIPTION
**Pull request by hfp**
* Register sparse matmul-op to use LIBXSMM's Bfloat16 functionality (functionality was there since ever but registration was lacking).
* Updated conv-integration with LIBXSMM (there is much more possible: https://github.com/hfp/libxsmm/issues/281).
* Generalized libxsmm.BUILD using glob/patterns.

This issue refers to https://github.com/hfp/libxsmm/issues/295

### CODE REVIEW

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @gunan: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @gunan: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by yifeif** [comment]
@jewillco do you mind taking a look at this PR? Thanks!

**Comment by hfp** [comment]
Is there any action needed from side to get this issue closed?

**Comment by penpornk** [comment]
@hfp I'm so sorry for our delay! I'll review this soon.

**Comment by hfp** [comment]
> @hfp I'm so sorry for our delay! I'll review this soon.

No worries!

I have updated the PR to the current TF/master. I also (temporarily) white-listed "libxsmm_archive" to allow a single URL for this dependency. When the PR is taken/squased, then (1) the change in `third_party/repo.bzl` can be simply omitted, (2) LIBXSMM archive to be mirrored, and (3) `tensorflow/workspace.bzl` updated according to the mirror.

File: tensorflow/workspace.bzl
Comment (penpornk) [inline tensorflow/workspace.bzl]: Please also add the mirror link. I'll add the file to the mirror later. (Bazel won't fail even if the link doesn't exist now.)
```
"http://mirror.tensorflow.org/github.com/hfp/libxsmm/archive/1ea106130554f252669015aa7613b1f05f42b723.zip
```

File: third_party/libxsmm.BUILD
Comment (penpornk) [inline third_party/libxsmm.BUILD]: You can just remove these if you don't think they are necessary anymore.

File: third_party/repo.bzl
Comment (penpornk) [inline third_party/repo.bzl]: Let's remove this line since we will have the mirror link.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you very much again for the PR! Your help is much appreciated. Also, sorry again for my lateness. Most of my comments are about styling. Formatting also seems off (lines too long, etc). Please apply `clang-format` to the PR.

Please go ahead and add the mirror URL. I'll make sure it is hosted soon. (Bazel won't fail even if the link doesn't work, so no harm there.)

This PR will probably take a while before it can be merged. We'll have to update our internal copy of libxsmm first, which can take some time.

File: tensorflow/workspace.bzl
Comment (penpornk) [reply tensorflow/workspace.bzl]: Update: I've put the file on the mirror. The mirror URL works now.

File: third_party/repo.bzl
Comment (hfp) [reply third_party/repo.bzl]: I will work on this next week, it's not forgotten. Let's take the time to finish it nicely.  I did not change the cluttered source code format so far to keep the patch small. I will happily pipe it through clang-format. Thank you for the review!

**Comment by rthadur** [comment]
@hfp can you please resolve conflicts.

**Comment by penpornk** [comment]
@hfp Thank you for letting us know! Please take your time. :) 
Good news: I don't think there's anything blocking the merge now.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by tensorflowbutler** [comment]
It has been 43 days that this pull-request has stalled. Please create a new pull-request with the requested changes.

**Comment by penpornk** [comment]
Hi @hfp,

Based on the comment in https://github.com/hfp/libxsmm/pull/355, I realized there might have been a misunderstanding.

First off, I apologize again for our delay in reviewing your PR. We are grateful for your valuable time and it wasn't our intention to make you feel neglected. We realized our response time has been very slow for many PRs and we have been trying to improve that. 

Secondly, I wanted to clarify that the PR has been closed strictly because of TF's PR policy. We close PRs if they have been inactive for longer than a certain period of time, just to make the number of active PRs we are monitoring manageable. This doesn't mean that we are dismissing the closed PRs, the authors are encouraged to reopen the PR when they are ready to make adjustments again. I think this is where the misunderstanding happened. I just realized that the PR status 'Closed/Rejected' is not very descriptive and that we could have explained the reasoning of automatically closing PRs better. I'm so sorry if this has caused you unnecessary frustrations.

I'm reopening the PR in case you'd still like to contribute. This PR was very close to being merged, it just needed minor formatting edits to conform with our C++ style guide. (We might need more changes now since TF code has also changed quite a bit in the past several months.) Please let me know if I can help with anything.

**Comment by hfp** [comment]
https://github.com/hfp/libxsmm/pull/355#issuecomment-546622659

**Comment by hfp** [comment]
I can update the PR to current TF/master. I think format changes should be a separate task. Perhaps otherwise the diff will be dominated with unrelated changes (I haven't done anything in this PR to make the format worse or non-conforming).

**Review by penpornk — DISMISSED** [review]
@hfp Thank you for verifying this with LIBXSMM 1.14 and sorry for my delay! (We have the TF World event last week.) 

Format changes: We might need some changes anyway (otherwise our internal checker won't accept the code). I'll take care of the edits internally to keep the changes minimal then. 

Could you please help resolve the merge conflicts?

**Comment by rthadur** [comment]
@hfp gentle ping to resolve conflicts.

**Comment by hfp** [comment]
I have force-pushed an update. Currently, tests are running.

**Review by penpornk — APPROVED** [review]
Thank you for the update!

**Pull request merged**
